nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -91,19 +95,187 @@ enum VolumeGetError {
+                            read_only_target,
+                        ),
+                    ));
+                }
+            }
+        }
+
+        Ok(volume)
+    }
+
+    async fn target_to_region(
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        err: &OptionalError<AddrParseError>,
+        target: &str,
+        read_only: bool,
+    ) -> Result<Option<Region>, diesel::result::Error> {
+        let address: SocketAddrV6 = target.parse().map_err(|e| err.bail(e))?;
+        let ip: db::model::Ipv6Addr = address.ip().into();
+
+        // Match region by dataset id and target port
+
+        use db::schema::dataset::dsl as dataset_dsl;
+        use db::schema::region::dsl as region_dsl;
+
+        dataset_dsl::dataset
+            .inner_join(
+                region_dsl::region
+                    .on(region_dsl::dataset_id.eq(dataset_dsl::id)),
+            )
+            .filter(dataset_dsl::ip.eq(ip))
+            .filter(
+                region_dsl::port
+                    .eq(Some::<db::model::SqlU16>(address.port().into())),
+            )
+            .filter(region_dsl::read_only.eq(read_only))
+            .select(Region::as_select())
+            .get_result_async::<Region>(conn)
+            .await
+            .optional()
+    }
+
+    async fn read_only_target_to_volume_resource_usage(
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        err: &OptionalError<AddrParseError>,
+        read_only_target: &str,
+    ) -> Result<Option<VolumeResourceUsage>, diesel::result::Error> {
+        // Easy case: it's a region snapshot, and we can match by the snapshot
+        // address directly
+
+        let maybe_region_snapshot = {
+            use db::schema::region_snapshot::dsl;
+            dsl::region_snapshot
+                .filter(dsl::snapshot_addr.eq(read_only_target.to_string()))
+                .filter(dsl::deleting.eq(false))
+                .select(RegionSnapshot::as_select())
+                .get_result_async::<RegionSnapshot>(conn)
+                .await
+                .optional()?
+        };
+
+        if let Some(region_snapshot) = maybe_region_snapshot {
+            return Ok(Some(VolumeResourceUsage::RegionSnapshot {
+                dataset_id: region_snapshot.dataset_id,
+                region_id: region_snapshot.region_id,
+                snapshot_id: region_snapshot.snapshot_id,
+            }));
         }
 
+        // Less easy case: it's a read-only region, and we have to match by
+        // dataset ip and region port
+
+        let maybe_region = Self::target_to_region(
+            conn,
+            err,
+            read_only_target,
+            true, // read-only
+        )
+        .await?;
+
+        if let Some(region) = maybe_region {
+            return Ok(Some(VolumeResourceUsage::ReadOnlyRegion {
+                region_id: region.id(),
+            }));
+        }
+
+        // If the resource was hard-deleted, or in the off chance that the
```

Comment:
It's not expected often, no - only during some race conditions where a resource was deleted before a volume that uses it could be created. In those cases we'll see `CouldNotFindResource` errors, so I'm inclined not to create a separate log message.